### PR TITLE
Check if htmlBodyId exists before trying to fetch htmlBody

### LINKED ISSUE #7646: IMAP Related Crash Stack Traces in NextCloud Cron Logs Every Night
### Steps to reproduce

Wait for cron.php error notification mails.

### Expected behavior

No error notification mail should appear.

### Actual behavior

Error notification mails appear twice a day (?) with the following stack trace:

```plain
Error: Call to a member function getBodyPart() on null in /var/www/nextcloud/apps/mail/lib/IMAP/MessageMapper.php:727
Stack trace:
#0 [internal function]: OCA\Mail\IMAP\MessageMapper->OCA\Mail\IMAP\{closure}()
#1 /var/www/nextcloud/apps/mail/lib/IMAP/MessageMapper.php(738): array_map()
#2 /var/www/nextcloud/apps/mail/lib/IMAP/PreviewEnhancer.php(86): OCA\Mail\IMAP\MessageMapper->getBodyStructureData()
#3 /var/www/nextcloud/apps/mail/lib/Service/PreprocessingService.php(81): OCA\Mail\IMAP\PreviewEnhancer->process()
#4 /var/www/nextcloud/apps/mail/lib/BackgroundJob/PreviewEnhancementProcessingJob.php(94): OCA\Mail\Service\PreprocessingService->process()
#5 /var/www/nextcloud/lib/public/BackgroundJob/Job.php(78): OCA\Mail\BackgroundJob\PreviewEnhancementProcessingJob->run()
#6 /var/www/nextcloud/lib/public/BackgroundJob/TimedJob.php(103): OCP\BackgroundJob\Job->start()
#7 /var/www/nextcloud/lib/public/BackgroundJob/TimedJob.php(93): OCP\BackgroundJob\TimedJob->start()
#8 /var/www/nextcloud/cron.php(152): OCP\BackgroundJob\TimedJob->execute()
#9 {main}
```

### Mail app version

2.1.1

### Mailserver or service

Dovecot

### Operating system

Debian 11 Bullseye

### PHP engine version

PHP 7.4

### Web server

Apache (supported)

### Database

MariaDB

### Additional info

Errors seem to have started occuring since the upgrade to NextCloud 25.0.1 and maybe a corresponding mail app update implicitly performed then.

Or it may just be a coincidence and maybe started slightly earlier - not sure.

### PR DESCRIPTION
**Pull request by miaulalala**
Fixes #7646

### CODE REVIEW

File: lib/IMAP/MessageMapper.php [hunk 1]
```diff
@@ -724,7 +724,7 @@ public function getBodyStructureData(Horde_Imap_Client_Socket $client,
 			]);
 			/** @var Horde_Imap_Client_Data_Fetch $part */
 			$part = $parts[$fetchData->getUid()];
-			$htmlBody = $part->getBodyPart($htmlBodyId);
+			$htmlBody = ($htmlBodyId !== null) ? $part->getBodyPart($htmlBodyId) : [];
```
Comment (ChristophWurst) [inline lib/IMAP/MessageMapper.php:727]: Why an empty array and not an empty string or null?

Comment (miaulalala) [reply lib/IMAP/MessageMapper.php:727]: Good question
image

**Comment by tobiasKaminsky** [comment]
Experience this also.
I am testing the fix right now.

**Comment by tobiasKaminsky** [comment]
After ~1h testing, with running cron every 5min, no problems.
But I have no idea about impact, and also had no problems with Mail, but I just saw this error.

**Comment by tobiasKaminsky** [comment]
Code is still in place, but I again receive:
```
Error: Call to a member function getBodyPart() on null in /srv/nextcloud/apps/mail/lib/IMAP/MessageMapper.php:727
Stack trace:
#0 [internal function]: OCA\Mail\IMAP\MessageMapper->OCA\Mail\IMAP\{closure}(Object(Horde_Imap_Client_Data_Fetch))
#1 /srv/nextcloud/apps/mail/lib/IMAP/MessageMapper.php(738): array_map(Object(Closure), Array)
#2 /srv/nextcloud/apps/mail/lib/IMAP/PreviewEnhancer.php(86): OCA\Mail\IMAP\MessageMapper->getBodyStructureData(Object(Horde_Imap_Client_Socket), 'INBOX', Array)
#3 /srv/nextcloud/apps/mail/lib/Service/PreprocessingService.php(81): OCA\Mail\IMAP\PreviewEnhancer->process(Object(OCA\Mail\Account), Object(OCA\Mail\Db\Mailbox), Array)
#4 /srv/nextcloud/apps/mail/lib/BackgroundJob/PreviewEnhancementProcessingJob.php(94): OCA\Mail\Service\PreprocessingService->process(1668459008, Object(OCA\Mail\Account))
#5 /srv/nextcloud/lib/public/BackgroundJob/Job.php(78): OCA\Mail\BackgroundJob\PreviewEnhancementProcessingJob->run(Array)
#6 /srv/nextcloud/lib/public/BackgroundJob/TimedJob.php(103): OCP\BackgroundJob\Job->start(Object(OC\BackgroundJob\JobList))
#7 /srv/nextcloud/lib/public/BackgroundJob/TimedJob.php(93): OCP\BackgroundJob\TimedJob->start(Object(OC\BackgroundJob\JobList))
#8 /srv/nextcloud/cron.php(152): OCP\BackgroundJob\TimedJob->execute(Object(OC\BackgroundJob\JobList), Object(OC\Log))
#9 {main}

```

**Comment by tobiasKaminsky** [comment]
Code is still in place, but I again receive:
```
Error: Call to a member function getBodyPart() on null in /srv/nextcloud/apps/mail/lib/IMAP/MessageMapper.php:727
Stack trace:
#0 [internal function]: OCA\Mail\IMAP\MessageMapper->OCA\Mail\IMAP\{closure}(Object(Horde_Imap_Client_Data_Fetch))
#1 /srv/nextcloud/apps/mail/lib/IMAP/MessageMapper.php(738): array_map(Object(Closure), Array)
#2 /srv/nextcloud/apps/mail/lib/IMAP/PreviewEnhancer.php(86): OCA\Mail\IMAP\MessageMapper->getBodyStructureData(Object(Horde_Imap_Client_Socket), 'INBOX', Array)
#3 /srv/nextcloud/apps/mail/lib/Service/PreprocessingService.php(81): OCA\Mail\IMAP\PreviewEnhancer->process(Object(OCA\Mail\Account), Object(OCA\Mail\Db\Mailbox), Array)
#4 /srv/nextcloud/apps/mail/lib/BackgroundJob/PreviewEnhancementProcessingJob.php(94): OCA\Mail\Service\PreprocessingService->process(1668459008, Object(OCA\Mail\Account))
#5 /srv/nextcloud/lib/public/BackgroundJob/Job.php(78): OCA\Mail\BackgroundJob\PreviewEnhancementProcessingJob->run(Array)
#6 /srv/nextcloud/lib/public/BackgroundJob/TimedJob.php(103): OCP\BackgroundJob\Job->start(Object(OC\BackgroundJob\JobList))
#7 /srv/nextcloud/lib/public/BackgroundJob/TimedJob.php(93): OCP\BackgroundJob\TimedJob->start(Object(OC\BackgroundJob\JobList))
#8 /srv/nextcloud/cron.php(152): OCP\BackgroundJob\TimedJob->execute(Object(OC\BackgroundJob\JobList), Object(OC\Log))
#9 {main}

```

**Comment by tobiasKaminsky** [comment]
It seems to occur only every 12h, thus my previous 1h testing and reporting was too short.

**Comment by miaulalala** [comment]
@tobiasKaminsky up for some more testing?

**Comment by tobiasKaminsky** [comment]
> @tobiasKaminsky up for some more testing?

Always. Shall I just apply latest state of this PR?

**Comment by tobiasKaminsky** [comment]
As written above my last error was at 28.12.
I still applied latest version of this fix, but I am not sure if it changed anything…

**Comment by jcdickinson** [comment]
I manually patched this into my installation, and it resolves being unable to use Mail with Mailfence.

**Review by st3iny — APPROVED** [review]

**Pull request merged**